File: src/main/standby.ts
import { getDb } from './db'
import { getTrackById } from './library'
import { StandbyEntry } from '@shared/types'

/** Ordered track ids in the standby playlist. */
export function standbyTrackIds(): number[] {
  return (
    getDb().prepare('SELECT track_id FROM standby ORDER BY position').all() as {
      track_id: number
    }[]
  ).map((r) => r.track_id)
}

export function listStandby(): StandbyEntry[] {
  const rows = getDb().prepare('SELECT id, track_id FROM standby ORDER BY position').all() as {
    id: number
    track_id: number
  }[]
  return rows
    .map((r) => {
      const track = getTrackById(r.track_id)
      return track ? { id: r.id, track } : null
    })
    .filter((e): e is StandbyEntry => e !== null)
}

export function addStandby(trackId: number): void {
  if (!getTrackById(trackId)) throw new Error('Track not found')
  const db = getDb()
  const pos = (db.prepare('SELECT COALESCE(MAX(position), 0) + 1 AS p FROM standby').get() as {
    p: number
  }).p
  db.prepare('INSERT INTO standby (track_id, position, added_at) VALUES (?, ?, ?)').run(
    trackId,
    pos,
    new Date().toISOString()
  )
}

export function removeStandby(id: number): void {
  getDb().prepare('DELETE FROM standby WHERE id = ?').run(id)
}

export function clearStandby(): void {
  getDb().prepare('DELETE FROM standby').run()
}

File: crates/jukebox-core/src/engine/standby.rs
//! The standby playlist: filler for when the guest queue is empty. Ports src/main/standby.ts.

use rusqlite::Connection;

use super::{EngineError, Result};
use crate::db::iso_now;
use crate::library::query::track_by_id;
use crate::types::StandbyEntry;

/// The playlist's track ids, in order.
pub(super) fn track_ids(db: &Connection) -> rusqlite::Result<Vec<i64>> {
    db.prepare("SELECT track_id FROM standby ORDER BY position")?
        .query_map([], |r| r.get(0))?
        .collect()
}

pub(super) fn list(db: &Connection) -> Result<Vec<StandbyEntry>> {
    let rows: Vec<(i64, i64)> = db
        .prepare("SELECT id, track_id FROM standby ORDER BY position")?
        .query_map([], |r| Ok((r.get(0)?, r.get(1)?)))?
        .collect::<rusqlite::Result<_>>()?;
    let mut entries = Vec::new();
    for (id, track_id) in rows {
        if let Some(track) = track_by_id(db, track_id)? {
            entries.push(StandbyEntry { id, track });
        }
    }
    Ok(entries)
}

pub(super) fn add(db: &Connection, track_id: i64) -> Result<()> {
    if track_by_id(db, track_id)?.is_none() {
        return Err(EngineError::rejected(400, "Track not found"));
    }
    let position: i64 = db.query_row(
        "SELECT COALESCE(MAX(position), 0) + 1 AS p FROM standby",
        [],
        |r| r.get(0),
    )?;
    db.execute(
        "INSERT INTO standby (track_id, position, added_at) VALUES (?, ?, ?)",
        (track_id, position, iso_now()),
    )?;
    Ok(())
}

pub(super) fn remove(db: &Connection, id: f64) -> Result<()> {
    db.execute("DELETE FROM standby WHERE id = ?", [id])?;
    Ok(())
}

pub(super) fn clear(db: &Connection) -> Result<()> {
    db.execute("DELETE FROM standby", [])?;
    Ok(())
}
